runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+}
+
+impl crate::Storage for Storage {
+    type Blob = Blob;
+
+    async fn open(&self, partition: &str, name: &[u8]) -> Result<Blob, Error> {
+        let _guard = self.lock.lock().map_err(|_| Error::LockGrabFailed)?;
+
+        // Construct the full path
+        let path = self.storage_directory.join(partition).join(hex(name));
+        let parent = path
+            .parent()
+            .ok_or_else(|| Error::BlobOpenFailed(partition.into(), hex(name)))?;
+
+        // Create the partition directory if it does not exist
+        fs::create_dir_all(parent).map_err(|_| Error::PartitionCreationFailed(partition.into()))?;
+
+        // Open the file in read-write mode, create if it does not exist
+        let file = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(&path)
+            .map_err(|_| Error::BlobOpenFailed(partition.into(), hex(name)))?;
+
+        // Get the file length
+        let len = file.metadata().map_err(|_| Error::ReadFailed)?.len();
+
+        Ok(Blob::new(partition.into(), name, file, len))
+    }
+
+    async fn remove(&self, partition: &str, name: Option<&[u8]>) -> Result<(), Error> {
+        let _guard = self.lock.lock().map_err(|_| Error::LockGrabFailed)?;
+
+        let path = self.storage_directory.join(partition);
+        if let Some(name) = name {
+            let blob_path = path.join(hex(name));
+            fs::remove_file(blob_path)
+                .map_err(|_| Error::BlobMissing(partition.into(), hex(name)))?;
+        } else {
+            fs::remove_dir_all(path).map_err(|_| Error::PartitionMissing(partition.into()))?;
+        }
+        Ok(())
+    }
+
+    async fn scan(&self, partition: &str) -> Result<Vec<Vec<u8>>, Error> {
+        let _guard = self.lock.lock().map_err(|_| Error::LockGrabFailed)?;
+
+        let path = self.storage_directory.join(partition);
+
+        let entries =
+            std::fs::read_dir(&path).map_err(|_| Error::PartitionMissing(partition.into()))?;
+
+        let mut blobs = Vec::new();
+        for entry in entries {
+            let entry = entry.map_err(|_| Error::ReadFailed)?;
+            let file_type = entry.file_type().map_err(|_| Error::ReadFailed)?;
+
+            if !file_type.is_file() {
+                return Err(Error::PartitionCorrupt(partition.into()));
+            }
+
+            if let Some(name) = entry.file_name().to_str() {
+                let name = from_hex(name).ok_or(Error::PartitionCorrupt(partition.into()))?;
+                blobs.push(name);
+            }
+        }
+
+        Ok(blobs)
+    }
+}
+
+#[derive(Clone)]
+pub struct Blob {
+    partition: String,
+    name: Vec<u8>,
+    // (underlying file, iouring, blob length)
+    file: Arc<Mutex<(File, IoUring, u64)>>,
+}
+
+impl Blob {
+    pub fn new(partition: String, name: &[u8], file: File, len: u64) -> Self {
+        let ring = IoUring::new(32).unwrap();
```

Comment:
Now that things are properly organized, I advocate for going for the "shared ring" approach in this PR.